### PR TITLE
IPUT: delay flush till remote memory is updated

### PR DESCRIPTION
**Pull request by hoopoepg**
- on some systems iput operation may delay delivering data
  from HCA to memory on remote side. to workaround this issue
  add non-blocking get operation to ensure that remote memory
  is updated

Signed-off-by: Sergey Oblomov <sergeyo@mellanox.com>

### CODE REVIEW

**Comment by ibm-ompi** [comment]
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/f7bb53544f987c4815cf6287b6f026e3

**Comment by ibm-ompi** [comment]
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/674e0925c47b647dfe4952f340f96a31

**Comment by yosefe** [comment]
@hoopoepg can this happen only with iput, or also with a sequence of non-strided put-s?

**Comment by hoopoepg** [comment]
@yosefe technically it may happen on all put operations, but iput split message by small chunks which cause PCI overload

File: oshmem/shmem/c/shmem_iput.c [hunk 1]
```diff
@@ -23,6 +23,25 @@
...
+#if OSHMEM_SPEC_COMPAT == 1
+#  define SHMEM_REMOTE_DELIVERY(ctx, type, target, tst, nelemes, pe)       \
+        do {                                                               \
+            static type _tmp;                                              \
+            rc = MCA_SPML_CALL(get_nb(ctx,                                 \
+                                      (void*)(target + nelemes - 1 * tst), \
+                                      element_size,                        \
+                                      (void*)(&_tmp),                      \
+                                      pe, NULL));                          \
+        }while(0)
```
Comment (dmitrygladkov) [inline oshmem/shmem/c/shmem_iput.c]: minor: wrong indentation

**Comment by sam6258** [comment]
I tested the fix on the original issue and it has gone away. Looks good to me.

**Comment by manjugv** [comment]
@hoopoepg Unless I'm missing something (from the code perspective), I don't think there is a need for this here. The semantics **does not** guarantee the completion on the remote memory, when it returns. It just guarantees that the source buffers are reusable. 

For PUTs, you need to ensure this on subsequent invocation of shmem_quiet / barrier.

**Comment by hoopoepg** [comment]
@manjugv right, this fix just adds non-blocking get to end of put operation which guarantee delivery on flush

**Pull request closed without merging**